text/2022-05-12-multi-cluster-support.md

```diff
@@ -0,0 +1,349 @@
+   -> If the chaos-mesh chart is not installed, and there is no chaos using this cluster
+      remove the finalizer
+   -> Apply the chaos and return
+-> If the finalizer doesn't exist, add a finalizer
+-> Helm list the installed release/chart in remote cluster
+   -> If the chaos-mesh chart is not installed, and the `RemoteCluster` itself is not being 
+      deleted, install the Chaos Mesh through helm and list the helm release again
+   -> If the chaos-mesh chart is installed, `Installed` conditions turn to true, else, turn to false
+   -> If the chaos-mesh chart is installed, but the config is different from the current merged config, run helm upgrade
+-> If all pods of target Chaos Mesh are running, `Ready` conditions turn to true, else, turn to false
+```
+
+### Q&A
+
+1. Why do we need to manage the installation of Chaos Mesh?
+
+   Consider a normal SaaS provider, which is a typical user of Chaos Mesh, will
+   need to automatically create Kubernetes clusters for his users. To use Chaos
+   Mesh, he will also need to install Chaos Mesh in the cluster which is created
+   automatically by the software. It’s inconvenient to deploy Chaos Mesh
+   manually in this situation.
+
+   Instead of every such user writing their own program to manage the
+   installation of Chaos Mesh, we could provide an official way to install /
+   upgrade / uninstall the Chaos Mesh.
+
+2. Why do we define a new API to manage the deployment, rather than a go package
+   or cli tools?
+
+   Because the k8s API is just intuitive and easy to integrate, and it’s only
+   one step further than a go package or cli tools.
+
+3. Why is the `RemoteCluster` resource cluster scoped?
+
+   Different *Chaos (e.g. NetworkChaos) will want to inject chaos in the same
+   cluster. If the `RemoteCluster` is namespace scoped, and we have to allow
+   selecting a remote cluster across the namespaces, then the namespace itself
+   is meaningless.
+
+4. Where will the controller exist?
+
+   In the chaos-controller-manager.
+
+5. What will it install?
+
+   A normal Chaos Mesh, with some controllers (like dashboard, workflow,
+   schedule) disabled.
+
+6. When will the cluster be upgraded?
+
+   As discussed with @iguoyr, automatically upgrading is a dangerous behavior.
+   The controller will never upgrade the remote cluster automatically, except
+   when the user add an annotation to tell us he want to upgrade the cluster. We
+   could also provide a global configuration to always upgrade the cluster
+   automatically.
+
+   If the remote cluster version is using an outdated version, all functions
+   except upgrade and uninstalltion (like modifying the configuration) will not
+   work.
+
+   Upgrading the cluster will also `replace` the CRD.
+
+## Inject
+
+To inject chaos in another cluster, add a `cluster` and `namespace` field to
+every selector. For example
+
+```yaml
+apiVersion: chaos-mesh.org/v1alpha1
+kind: StressChaos
+metadata:
+ name: burn-cpu
+spec:
+ remoteCluster:
```

Comment:
Can `remoteCluster` be used as a selector instead of a single field?

Reply:
Indeed, we cannot assume the different clusters are isomorphic, so if you are trying to run the same selector on different remote cluster, I doubt whether it's meaningful. Just like the `containerNames`, which is also embarrassing, as its existence shows our assumption of the isomorphic between pods.

And it's also hard to manage the status for chaos in multiple different clusters.